Avoid un-necessary unwrapping of Tensor in SavedVariable

Varal7 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59838 Remove dead code in SavedVariable
* **#59837 Avoid un-necessary unwrapping of Tensor in SavedVariable**

Summary: Fixes #58500

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29069215](https://our.internmc.facebook.com/intern/diff/D29069215)

Varal7 committed:
Avoid un-necessary unwrapping of Tensor in SavedVariable

Summary: Fixes #58500

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Varal7 committed:
Update on "Avoid un-necessary unwrapping of Tensor in SavedVariable"


Summary: Fixes #58500

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

Varal7 committed:
Update on "Avoid un-necessary unwrapping of Tensor in SavedVariable"


Summary: Fixes #58500

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/saved_variable.cpp:
```diff
@@ -41,6 +41,21 @@ SavedVariable::SavedVariable(const Variable& variable, bool is_output, bool is_i
     version_counter_ = impl::version_counter(variable);
     saved_version_ = version_counter_.current_version();
 
+    // This check actually does not hold
```
Why? When you consider inplace ops?

albanD commented on torch/csrc/autograd/saved_variable.cpp:
```diff
@@ -41,6 +41,21 @@ SavedVariable::SavedVariable(const Variable& variable, bool is_output, bool is_i
+    // If the variable is not an output, its grad_fn has already been fully
+    // created and in particular will be a different Node than the one
+    // we are currently constructing (the one that owns this SavedVariable).
+    // Similarly, if the variable is a leaf.
```
I think the leaf case is different actually. You don't have a cycle because the Tensor only has a weak ref to the grad_accumulator. So that cannot create a cycle.

albanD commented on torch/csrc/autograd/saved_variable.h:
```diff
@@ -70,5 +74,6 @@ class TORCH_API SavedVariable {
   bool requires_grad_ = false;
   bool has_grad_fn_ = false;
   bool is_inplace_view_ = false;
+  bool saved_original = false;
```
Use `_` at the end for non-public attributes

albanD commented on torch/csrc/autograd/saved_variable.h:
```diff
@@ -46,6 +46,10 @@ class TORCH_API SavedVariable {
   }
 
  private:
+  // The variable to save.
```
The doc here should mention the link between the content of this and the boolean saved_original below.

albanD commented on torch/csrc/autograd/saved_variable.cpp:
```diff
@@ -41,6 +41,21 @@ SavedVariable::SavedVariable(const Variable& variable, bool is_output, bool is_i
     version_counter_ = impl::version_counter(variable);
```
Why do you still save the version counter here?

Varal7 replied:
Yes I think you're right that I don't need it here.

albanD commented on test/test_autograd.py:
```diff
@@ -5421,6 +5421,26 @@ def get_refs(with_backward):
+        with self.assertRaisesRegex(RuntimeError, "after they have already been freed"):
+            c.grad_fn._saved_self
+
+        self.assertEqual(a, torch.ones(5))
```
Why do you check the value of a here? Did you meant to check the value of a.grad? If so, maybe use another value than 1 for a so that you get non-1 gradients.

Varal7 replied:
I meant to check that the value of `a` has not been cleared by `release_variables`.

albanD replied:
Ah! Maybe leave a comment there "a is still untouched"?
Also, unless we do something very bad with refcounting, there is no reason for a to have disapeared here.

albanD commented on test/test_autograd.py:
```diff
@@ -7332,7 +7352,9 @@ def test_sparse_backward(self, device, dtype):
             @staticmethod
             def forward(ctx, x, grad_x):
-                ctx.save_for_backward(grad_x)
+                # this is needed because of https://github.com/pytorch/pytorch/issues/59828
```
The fix landed now. So after rebasing on top of master, you should be able to remove this.

albanD reviewed (commented):
Looks ok to me.
Is there any other tests that we want to do here?

Varal7 commented:
> Looks ok to me.
> Is there any other tests that we want to do here?

Thanks for the review! I will address your comments inline.
I'm happy with the current tests (after I fix the one you pointed out)

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29069215).

Varal7 committed:
Update on "Avoid un-necessary unwrapping of Tensor in SavedVariable"


Summary: Fixes #58500

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29069215](https://our.internmc.facebook.com/intern/diff/D29069215)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29069215).

albanD commented on test/test_autograd.py:
```diff
@@ -5467,6 +5467,28 @@ def test_input_buffer_accum(self):
+        a = torch.randn(5, requires_grad=True)
+        a_orig = a.detach().clone()
+        b = a * a
+        c = a * b
```
nit: if you want to test an output here to make sure it is properly unwrap, you can use exp() and make sure the `_saved_result` is not the same as the returned Tensor.

albanD commented on torch/csrc/autograd/saved_variable.cpp:
```diff
@@ -37,13 +37,30 @@ SavedVariable::SavedVariable(const Variable& variable, bool is_output, bool is_i
-    version_counter_ = impl::version_counter(variable);
-    saved_version_ = version_counter_.current_version();
+    is_inplace_on_view_ = is_inplace_on_view;
+    auto& version_counter = impl::version_counter(variable);
```
Why not `const auto&` ?

albanD commented on torch/csrc/autograd/saved_variable.cpp:
```diff
@@ -37,13 +37,30 @@ SavedVariable::SavedVariable(const Variable& variable, bool is_output, bool is_i
+    saved_version_ = version_counter.current_version();
+
+    // This check actually does not hold
+    // TORCH_CHECK(is_output && variable.is_leaf(), "Variable is both an output and a leaf");
```
This commented code should be removed.

albanD commented on torch/csrc/autograd/saved_variable.h:
```diff
@@ -46,6 +46,15 @@ class TORCH_API SavedVariable {
+  // 2. or its tensor_data.
+  // If storing the variable itself would create a circular reference,
+  // we fall into the second case and its metadata is also saved separately.
+  // In that case, the grad_fn must be passed in to the unpack function when
```
nit: not true for inplace_on_view case.

albanD reviewed (approved):
LGTM
Minor updates to remove spurious code.

Varal7 committed:
Update on "Avoid un-necessary unwrapping of Tensor in SavedVariable"


Summary: Fixes #58500

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29069215](https://our.internmc.facebook.com/intern/diff/D29069215)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29069215).

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29069215).
